py2puml/parsing/astvisitors.py

```diff
@@ -0,0 +1,208 @@
+            self.self_attributes.append(Variable(node.attr, self.annotation))
+
+    def visit_Subscript(self, node: Subscript):
+        '''
+        Assigns a value to a subscript of an existing variable: must be skipped
+        '''
+        pass
+
+
+class ConstructorVisitor(NodeVisitor):
+    '''Identifies the attributes (and infer their type) assigned to self in the body of a constructor method'''
+    def __init__(self, constructor_source: str, class_name: str, root_fqdn: str, module_resolver: ModuleResolver, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.constructor_source = constructor_source
+        self.class_fqdn: str = f'{module_resolver.module.__name__}.{class_name}'
+        self.root_fqdn = root_fqdn
+        self.module_resolver = module_resolver
+        self.class_self_id: str = None
+        self.variables_namespace: List[Variable] = []
+        self.uml_attributes: List[UmlAttribute] = []
+        self.uml_relations_by_target_fqdn: Dict[str, UmlRelation] = {}
+
+    def extend_relations(self, target_fqdns: List[str]):
+        self.uml_relations_by_target_fqdn.update({
+            target_fqdn: UmlRelation(self.class_fqdn, target_fqdn, RelType.COMPOSITION)
+            for target_fqdn in target_fqdns
+            if target_fqdn.startswith(self.root_fqdn) and target_fqdn not in self.uml_relations_by_target_fqdn
+        })
+
+    def is_in_variables_namespace(self, variable_id: str) -> bool:
+        return self.get_from_namespace(variable_id) != None
+
+    def get_from_namespace(self, variable_id: str) -> Variable:
+        return next((
+            variable
+            # variables namespace is iterated antichronologically
+            # to account for variables being overridden
+            for variable in self.variables_namespace[::-1]
+            if variable.id == variable_id
+        ), None)
+
+    def generic_visit(self, node):
+        NodeVisitor.generic_visit(self, node)
+
+    def visit_FunctionDef(self, node: FunctionDef):
+        # retrieves constructor arguments ('self' reference and typed arguments)
+        if node.name == '__init__':
+            variables_collector = SignatureVariablesCollector(self.constructor_source)
+            variables_collector.visit(node)
+            self.class_self_id: str = variables_collector.class_self_id
+            self.variables_namespace = variables_collector.variables
+
+        self.generic_visit(node)
+
+    def visit_AnnAssign(self, node: AnnAssign):
+        variables_collector = AssignedVariablesCollector(self.class_self_id, node.annotation)
+        variables_collector.visit(node.target)
+
+        short_type, full_namespaced_definitions = self.derive_type_annotation_details(node.annotation)
+        # if any, there is at most one self-assignment
+        for variable in variables_collector.self_attributes:
+            self.uml_attributes.append(UmlAttribute(variable.id, short_type, False))
```

Comment:
This suggestions also applies to other places in the code.
```suggestion
            self.uml_attributes.append(UmlAttribute(variable.id, short_type, static=False))
```

Reply:
thank you @jjerphan for your review, I will take a look at it this week.
Have a nice day

Reply:
changed